kafka-client/src/main/java/dev/responsive/kafka/api/async/internals/AsyncProcessor.java

```diff
@@ -364,42 +376,55 @@ public void flushAndAwaitPendingEvents() {
-      // runs if there are currently slow events being handled by the async threads.
-      checkUncaughtExceptionsOnAsyncThreads();
-
-      // Start by scheduling all unblocked events to hand off any events that
-      // were just unblocked by whatever we just finalized
-      final int numScheduled = drainSchedulingQueue();
-
-      // Need to finalize at least one event per iteration, otherwise there's no
-      // point returning to the scheduling queue since nothing new was unblocked
-      final int numFinalized = finalizeAtLeastOneEvent();
-      log.debug("Scheduled {} events and finalized {} events",
-                numScheduled, numFinalized
-      );
-    }
+    try {
+      // Make a (non-blocking) pass through the finalizing queue up front, to
+      // free up any recently-processed events before we attempt to drain the
+      // scheduling queue
+      drainFinalizingQueue();
+
+      while (!isCleared()) {
+        // Start by scheduling all unblocked events to hand off any events that
+        // were just unblocked by whatever we just finalized
+        final int numScheduled = drainSchedulingQueue();
+
+        // Need to finalize at least one event per iteration, otherwise there's no
+        // point returning to the scheduling queue since nothing new was unblocked
+        final int numFinalized = finalizeAtLeastOneEvent();
+        log.trace("Scheduled {} events and finalized {} events",
+                  numScheduled, numFinalized
+        );
+      }
 
-    assertQueuesEmpty();
-  }
+      assertQueuesEmpty();
 
-  private void checkUncaughtExceptionsOnAsyncThreads() {
-    final var uncaught = threadPool.checkUncaughtExceptions(asyncProcessorName, taskId.partition());
-    if (uncaught.isPresent()) {
-      if (fatalException == null) {
-        fatalException = uncaught.map(t ->
-          t instanceof RuntimeException ? (RuntimeException) t : new RuntimeException(t)).get();
-      }
+    } catch (final StreamsException streamsException) {
+      fatalException = streamsException.taskId().isPresent()
+        ? streamsException
+        : new StreamsException(streamsException, taskId);
+      throw fatalException;
+    } catch (final Throwable t) {
+      fatalException = new StreamsException(t, taskId);
       throw fatalException;
     }
   }
 
+  /**
+   * Check the AsyncThreadPool for any uncaught exceptions that arose while processing
+   * events or other fatal errors in the AsyncThreadPool itself.
+   * Note: since it's important to fail as quickly as possible in the event of a
+   * fatal exception, we don't limit the scope of this check to only errors
+   * tied to events that belong to this specific AsyncProcessor or task.
+   */
+  private void checkUncaughtExceptionsInAsyncThreadPool() throws Throwable {
+    final Optional<Throwable> fatal = threadPool.checkUncaughtExceptions(
+        asyncProcessorName, taskId.partition()
+    );
+    
+    if (fatal.isPresent()) {
```

Comment:
can we wrap it in a `RuntimeException` if its a checked exception? will help get rid of the `throws Throwable` everywhere which I don't think is adding anything

Reply:
Fair enough